From 3e8a31c256d9efdc6e46f5f58c1dd5e98bccd2ad Mon Sep 17 00:00:00 2001
From: David Ward <david.ward@ll.mit.edu>
Date: Wed, 13 May 2026 21:43:59 -0400
Subject: [PATCH 4/5] lib: northbound: Use lyd_new_term() in generic getter
 functions

lyd_new_term_bin() was removed in libyang 5.1.3. Modify the generic
getter functions in FRR's northbound API to use lyd_new_term().

Prior to this, it was assumed nb_oper_uint{64,32}_get() could be used
with nodes of different integer types - although nb_oper_uint64_get()
was only used with uint64 nodes, and nb_oper_uint32_get() was unused
(and contained an invalid pointer cast). Now, those functions handle
uint{64,32} nodes only and return a validation error for other nodes.
Introduce separate functions to handle the other types.

Signed-off-by: David Ward <david.ward@ll.mit.edu>
Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>
---
 lib/northbound.h      |  18 +++++
 lib/northbound_oper.c | 177 ++++++++++++++++++++++++++++--------------
 2 files changed, 138 insertions(+), 57 deletions(-)

diff --git a/lib/northbound.h b/lib/northbound.h
index f48cc34e3e..f1d10702d9 100644
--- a/lib/northbound.h
+++ b/lib/northbound.h
@@ -1610,6 +1610,24 @@ extern void *nb_oper_walk_finish_arg(void *walk);
 extern void *nb_oper_walk_cb_arg(void *walk);
 
 /* Generic getter functions */
+extern enum nb_error nb_oper_int8_get(const struct nb_node *nb_node,
+				      const void *parent_list_entry, struct lyd_node *parent);
+
+extern enum nb_error nb_oper_int16_get(const struct nb_node *nb_node,
+				       const void *parent_list_entry, struct lyd_node *parent);
+
+extern enum nb_error nb_oper_int32_get(const struct nb_node *nb_node,
+				       const void *parent_list_entry, struct lyd_node *parent);
+
+extern enum nb_error nb_oper_int64_get(const struct nb_node *nb_node,
+				       const void *parent_list_entry, struct lyd_node *parent);
+
+extern enum nb_error nb_oper_uint8_get(const struct nb_node *nb_node,
+				       const void *parent_list_entry, struct lyd_node *parent);
+
+extern enum nb_error nb_oper_uint16_get(const struct nb_node *nb_node,
+					const void *parent_list_entry, struct lyd_node *parent);
+
 extern enum nb_error nb_oper_uint32_get(const struct nb_node *nb_node,
 					const void *parent_list_entry, struct lyd_node *parent);
 
diff --git a/lib/northbound_oper.c b/lib/northbound_oper.c
index f10f112596..8481ec2f21 100644
--- a/lib/northbound_oper.c
+++ b/lib/northbound_oper.c
@@ -2345,69 +2345,113 @@ enum nb_error nb_oper_iterate_legacy(const char *xpath,
 	return ret;
 }
 
-static const char *_adjust_ptr(struct lysc_node_leaf *lsnode, const char *valuep, size_t *size)
+enum nb_error nb_oper_int8_get(const struct nb_node *nb_node, const void *parent_list_entry,
+			       struct lyd_node *parent)
 {
-	switch (lsnode->type->basetype) {
-	case LY_TYPE_INT8:
-	case LY_TYPE_UINT8:
-#if BYTE_ORDER == BIG_ENDIAN
-		valuep += 7;
-#endif
-		*size = 1;
-		break;
-	case LY_TYPE_INT16:
-	case LY_TYPE_UINT16:
-#if BYTE_ORDER == BIG_ENDIAN
-		valuep += 6;
-#endif
-		*size = 2;
-		break;
-	case LY_TYPE_INT32:
-	case LY_TYPE_UINT32:
-#if BYTE_ORDER == BIG_ENDIAN
-		valuep += 4;
-#endif
-		*size = 4;
-		break;
-	case LY_TYPE_INT64:
-	case LY_TYPE_UINT64:
-		*size = 8;
-		break;
-	case LY_TYPE_UNKNOWN:
-	case LY_TYPE_BINARY:
-	case LY_TYPE_STRING:
-	case LY_TYPE_BITS:
-	case LY_TYPE_BOOL:
-	case LY_TYPE_DEC64:
-	case LY_TYPE_EMPTY:
-	case LY_TYPE_ENUM:
-	case LY_TYPE_IDENT:
-	case LY_TYPE_INST:
-	case LY_TYPE_LEAFREF:
-	case LY_TYPE_UNION:
-	default:
-		assert(0);
-	}
-	return valuep;
+	struct lysc_node_leaf *lsnode = (struct lysc_node_leaf *)nb_node->snode;
+	struct lysc_node *snode = &lsnode->node;
+	ssize_t offset = (ssize_t)nb_node->cbs.get_elem;
+	void *valuep = (char *)parent_list_entry + offset;
+	char value[5]; /* INT8_MIN has a sign and then 3 digits */
+
+	if (lsnode->type->basetype != LY_TYPE_INT8)
+		return NB_ERR_VALIDATION;
+
+	snprintfrr(value, sizeof(value), "%" PRId8, *(int8_t *)valuep);
+	if (lyd_new_term(parent, snode->module, snode->name, value, LYD_NEW_PATH_UPDATE, NULL))
+		return NB_ERR_RESOURCE;
+	return NB_OK;
 }
 
-enum nb_error nb_oper_uint64_get(const struct nb_node *nb_node, const void *parent_list_entry,
-				 struct lyd_node *parent)
+enum nb_error nb_oper_int16_get(const struct nb_node *nb_node, const void *parent_list_entry,
+				struct lyd_node *parent)
+{
+	struct lysc_node_leaf *lsnode = (struct lysc_node_leaf *)nb_node->snode;
+	struct lysc_node *snode = &lsnode->node;
+	ssize_t offset = (ssize_t)nb_node->cbs.get_elem;
+	void *valuep = (char *)parent_list_entry + offset;
+	char value[7]; /* INT16_MIN has a sign and then 5 digits */
+
+	if (lsnode->type->basetype != LY_TYPE_INT16)
+		return NB_ERR_VALIDATION;
+
+	snprintfrr(value, sizeof(value), "%" PRId16, *(int16_t *)valuep);
+	if (lyd_new_term(parent, snode->module, snode->name, value, LYD_NEW_PATH_UPDATE, NULL))
+		return NB_ERR_RESOURCE;
+	return NB_OK;
+}
+
+enum nb_error nb_oper_int32_get(const struct nb_node *nb_node, const void *parent_list_entry,
+				struct lyd_node *parent)
+{
+	struct lysc_node_leaf *lsnode = (struct lysc_node_leaf *)nb_node->snode;
+	struct lysc_node *snode = &lsnode->node;
+	ssize_t offset = (ssize_t)nb_node->cbs.get_elem;
+	void *valuep = (char *)parent_list_entry + offset;
+	char value[12]; /* INT32_MIN has a sign and then 10 digits */
+
+	if (lsnode->type->basetype != LY_TYPE_INT32)
+		return NB_ERR_VALIDATION;
+
+	snprintfrr(value, sizeof(value), "%" PRId32, *(int32_t *)valuep);
+	if (lyd_new_term(parent, snode->module, snode->name, value, LYD_NEW_PATH_UPDATE, NULL))
+		return NB_ERR_RESOURCE;
+	return NB_OK;
+}
+
+enum nb_error nb_oper_int64_get(const struct nb_node *nb_node, const void *parent_list_entry,
+				struct lyd_node *parent)
+{
+	struct lysc_node_leaf *lsnode = (struct lysc_node_leaf *)nb_node->snode;
+	struct lysc_node *snode = &lsnode->node;
+	ssize_t offset = (ssize_t)nb_node->cbs.get_elem;
+	void *valuep = (char *)parent_list_entry + offset;
+	char value[22]; /* INT64_MIN has a sign and then 20 digits */
+
+	if (lsnode->type->basetype != LY_TYPE_INT64)
+		return NB_ERR_VALIDATION;
+
+	snprintfrr(value, sizeof(value), "%" PRId64, *(int64_t *)valuep);
+	if (lyd_new_term(parent, snode->module, snode->name, value, LYD_NEW_PATH_UPDATE, NULL))
+		return NB_ERR_RESOURCE;
+	return NB_OK;
+}
+
+enum nb_error nb_oper_uint8_get(const struct nb_node *nb_node, const void *parent_list_entry,
+				struct lyd_node *parent)
 {
 	struct lysc_node_leaf *lsnode = (struct lysc_node_leaf *)nb_node->snode;
 	struct lysc_node *snode = &lsnode->node;
 	ssize_t offset = (ssize_t)nb_node->cbs.get_elem;
-	uint64_t ubigval = *(uint64_t *)((char *)parent_list_entry + offset);
-	const char *valuep;
-	size_t size;
+	void *valuep = (char *)parent_list_entry + offset;
+	char value[4]; /* UINT8_MAX has 3 digits */
 
-	valuep = _adjust_ptr(lsnode, (const char *)&ubigval, &size);
-	if (yang_new_term_bin(parent, snode->module, snode->name, valuep, size,
-			      LYD_NEW_PATH_UPDATE, NULL))
+	if (lsnode->type->basetype != LY_TYPE_UINT8)
+		return NB_ERR_VALIDATION;
+
+	snprintfrr(value, sizeof(value), "%" PRIu8, *(uint8_t *)valuep);
+	if (lyd_new_term(parent, snode->module, snode->name, value, LYD_NEW_PATH_UPDATE, NULL))
 		return NB_ERR_RESOURCE;
 	return NB_OK;
 }
 
+enum nb_error nb_oper_uint16_get(const struct nb_node *nb_node, const void *parent_list_entry,
+				 struct lyd_node *parent)
+{
+	struct lysc_node_leaf *lsnode = (struct lysc_node_leaf *)nb_node->snode;
+	struct lysc_node *snode = &lsnode->node;
+	ssize_t offset = (ssize_t)nb_node->cbs.get_elem;
+	void *valuep = (char *)parent_list_entry + offset;
+	char value[6]; /* UINT16_MAX has 5 digits */
+
+	if (lsnode->type->basetype != LY_TYPE_UINT16)
+		return NB_ERR_VALIDATION;
+
+	snprintfrr(value, sizeof(value), "%" PRIu16, *(uint16_t *)valuep);
+	if (lyd_new_term(parent, snode->module, snode->name, value, LYD_NEW_PATH_UPDATE, NULL))
+		return NB_ERR_RESOURCE;
+	return NB_OK;
+}
 
 enum nb_error nb_oper_uint32_get(const struct nb_node *nb_node, const void *parent_list_entry,
 				 struct lyd_node *parent)
@@ -2415,13 +2459,32 @@ enum nb_error nb_oper_uint32_get(const struct nb_node *nb_node, const void *pare
 	struct lysc_node_leaf *lsnode = (struct lysc_node_leaf *)nb_node->snode;
 	struct lysc_node *snode = &lsnode->node;
 	ssize_t offset = (ssize_t)nb_node->cbs.get_elem;
-	uint64_t ubigval = *(uint64_t *)((char *)parent_list_entry + offset);
-	const char *valuep;
-	size_t size;
+	void *valuep = (char *)parent_list_entry + offset;
+	char value[11]; /* UINT32_MAX has 10 digits */
+
+	if (lsnode->type->basetype != LY_TYPE_UINT32)
+		return NB_ERR_VALIDATION;
+
+	snprintfrr(value, sizeof(value), "%" PRIu32, *(uint32_t *)valuep);
+	if (lyd_new_term(parent, snode->module, snode->name, value, LYD_NEW_PATH_UPDATE, NULL))
+		return NB_ERR_RESOURCE;
+	return NB_OK;
+}
+
+enum nb_error nb_oper_uint64_get(const struct nb_node *nb_node, const void *parent_list_entry,
+				 struct lyd_node *parent)
+{
+	struct lysc_node_leaf *lsnode = (struct lysc_node_leaf *)nb_node->snode;
+	struct lysc_node *snode = &lsnode->node;
+	ssize_t offset = (ssize_t)nb_node->cbs.get_elem;
+	void *valuep = (char *)parent_list_entry + offset;
+	char value[21]; /* UINT64_MAX has 20 digits */
+
+	if (lsnode->type->basetype != LY_TYPE_UINT64)
+		return NB_ERR_VALIDATION;
 
-	valuep = _adjust_ptr(lsnode, (const char *)&ubigval, &size);
-	if (yang_new_term_bin(parent, snode->module, snode->name, valuep, size,
-			      LYD_NEW_PATH_UPDATE, NULL))
+	snprintfrr(value, sizeof(value), "%" PRIu64, *(uint64_t *)valuep);
+	if (lyd_new_term(parent, snode->module, snode->name, value, LYD_NEW_PATH_UPDATE, NULL))
 		return NB_ERR_RESOURCE;
 	return NB_OK;
 }
-- 
2.55.0

